### ISSUE TITLE
Allow Noether staking pool operators to increase fees?

### ISSUE THREAD

**Issue by fcecin** [cartesi/cips#1]
This is the first discussion thread in this forum, which currently is the primary forum for documenting all Cartesi technical governance discussions that could potentially go through the Cartesi Improvement Proposal (CIP) process. 

This post will be a bit longer than necessary for kicking off the governance topic at hand (allowing Noether pools to increase fees) because I want to take this as an opportunity to also carry out some "meta" discussions about the CIP process itself, and specifically, about the role of this technical governance forum (I am currently the CIP Editor -- a role that is formalized in the first and current version of the CIP-1).

I have read the community discussions on Discord about the specific topic of allowing pool operators to increase their fees. Although Discord is great for discussions of the "chat" type, the chat logs of a Discord server are not really amenable for referencing within a formal governance process. For that, it is preferable that we use a communication tool that is like a web forum, where discussion pages can be referenced publicly with a static URL.

So, for now, Cartesi has decided to use GitHub issues in the CIP repository as discussion threads that can be referenced within the formal CIP technical standards documents. We may move this functionality to a dedicated forum operated by the Cartesi project in the future, if we deem it necessary. If we ever decide to do so, this will be communicated through a discussion thread and a Meta CIP here in this repository, which the community will have adequate time to debate, as with all CIPs.

The current workflow of community influence into Cartesi's technical direction is through a direct interaction with the entire team through channels like Discord. That works great as the sole channel for the start of a project, and it will be a path for governance for a long time to come, but, as a community, involving all stakeholders, we need to start moving towards more formal processes, especially as a project like Cartesi starts to grow in all sizes and scopes at a breakneck pace.

Any member of the community could have summarized the initial Discord discussions on this topic here. Currently, the member of the community that's seemingly most interested in bootstrapping the gradual formalization process of requests that are technical in nature is Cartesi itself. Hence, I, the CIP Editor, have been drafted to inaugurate and demonstrate this governance forum through this post. Which is why this post is heavily meta. But let's move into the actual technical topic.

In a nutshell, each staking pool has an operator's fee collected by the operator. The system would, arguably, be perfectly workable if that fee was entirely immutable, that is, the fee schedule is a constant attribute that is set upon the creation of the staking pool. Then, if an operator wants to change the fee, they can just create another staking pool, and invite the clients of their previous pool to join the new pool, or merely extinguish the old pool.

Now, having the fee be a changeable attribute is helpful, because if, for whatever reasons, the operator wants to change that fee, they and their clients will have a lot less work: the operator changes the fee and that is that. However, adjusting a pool fee upwards creates a window for abuse. If there is no reliable way to notify all of the (human) clients of a staking pool that the fee has been jacked up, they may be automatically opted out of being able to select another pool with more competitive fees.

But adjusting the pool fee downwards does not present that dilemma, which prompted Cartesi to take the free technical upside of allowing pool operators to decrease their fees. If operators want to increase them, they would still be required to create another pool with the higher fee, and convince the clients of the previous pool to migrate.

The community has been signaling an interest to actually allow pool operators to increase their fees in some way through Discord, and some elements of design for this feature have been presented. This governance thread is an invitation for a debate on the pros and cons of allowing fee increases in general, and, more importantly, an invitation for the community, including Cartesi itself, to sketch a technical design for a specific change that would conform to the idea of "allowing fee increases", i.e. by how much exactly, and how that would actually work -- would it work exactly how pool stakers are allowed to decrease fees, or would it be a more restricted mechanism? If those discussions have already been made on the Discord channel, they can be simply ported (copied, edited) into this GitHub issue, with a reference to the original author(s) of the designs and arguments when possible.

Possibly, this discussion here will evolve to a point where an actual Core CIP will be drafted by a member (any member, including Cartesi itself) of the community, proposing a specified implementation change, which will act as the CIPs owner. If that owner is not an employee of the Cartesi Project, still, probably, that member of the community will get assistance from us to get that CIP written and pushed successfully through the process. Another possible outcome is that the conclusion is that the current system is sufficient, and the resulting discussion in this forum can be forwarded to anyone who raises the topic in the future, meaning they have to consider all of the previous wealth of technical governance discussion that has already been accumulated on the topic in addressable form.

**Comment by cf-cartesi** [comment]
Currently fees can be defined at pool creation time and lowered afterwards,  but not raised. There are 2 commission models: flat rate and gas-based. On flat rate the defined % of the commission is taken by the pool whenever a block is produced. It is a simple model that might increase or decrease a lot the rewards of the pool owner depending on the gas/ETH/CTSI relation. The gas-based commission is more complex: the pool owner defines a gas cost per block produced and the commission is calculated on-the-fly using a chainlink oracle and uniswap to convert the gas value to the corresponding amount of CTSI to reward the pool owner. This model is more complex and harder to explain to users, but should provide a better fit to variations on the gas/ETH/CTSI relation.

Some community members raised a concern about not being able to raise the fees with the argument that it would hinder competition. The rationale is that a new pool could offer lower rates when starting to draw users, and once it reaches critical mass, could raise the fee to a value that would make it worth for the pool owner to maintain. It makes sense, but we should have some limitations on how to do that to prevent abuses and allow users time to opt-out and change pools if they wish to do so. One idea is having either a maximum variation on the commission value (could be absolute or relative) and a cool-down time to make a new change. No matter the model, some things should be taken into consideration besides the interest of pool users and pool owners: the more complex the logic, the more gas it will spend and that is bad for all parties involved; more complex logic is also more prone to leading to unexpected conditions which translate into bugs and security issues, so that things should be kept as simple as possible to avoid that; the models should also be simple for the users to understand and align their expectations, avoiding frustration.

**Comment by packetstracer** [comment]
Here Feida from Blockscope. As I stated in the discord discussions, I like the idea of being able to change the commission because it's good for competitiveness and in the end for decentralization, and even more with having a high block generation time.

A low fee can attract delegators and allow validators to grow and mine blocks in a decent interval for the delegators. Once achieved this threshold they can adjust the fee to still being competitive but sustainable. With a regular fee, small validators won't be able to compete with big validators, because users will prefer lower mining intervals.

Another issue would great variability in ETH gas prices, we've seen in the past how the ETH gas price has been incremented by 10-20x, and so if the validator set a quite low fee, could incur in loses and may do the task less sustainable.

**Comment by vaddec-everstake** [comment]
In my opinion, you need to be able to increase the commission for your pool. But for some certain percentage in (epoch / week / month / quarter). So that only once during this period this transaction can be made. And also, information is required that the commission for this pool has increased if this increase occurred in (epoch / week / month / quarter).

**Comment by augustoteixeira** [comment]
It is important to understand two aspects here:
- Cartesi's commission system already accounts for increases in Ethereum fees. If said fees go high, then the commission increases as well
- It has always been possible to decrease the commission, what is at question is whether it should be possible to increase them.
- Note that some users may subscribe to the pool and never look at this again. So if a pool increases its commission, the users may be loosing money without realizing.
- Also it is important to observe that the users pay a reasonable cost to swap between two different pools. So if a pool with lots of users decides to increase its commission, the users may become "trapped" in a situation that is not clear whether it is worth paying the transaction cost of changing pools to look for some other group (which can also increase the fees later on).

**Comment by fcecin** [comment]
If I were to design a pool fee increase feature, I would avoid any complicated time-based mechanisms. Instead, I would allow, during pool creation, a *range* of fees to be specified. So, when users sign up to a pool, they can choose pools that narrow down that fee range to a point, to e.g. a single scalar % value, if they want absolute certainty of what the fee is. Otherwise, pool operators can change the fee freely within the preset and immutable range, without complicated time locks, rules, restrictions, etc. The operators can even limit themselves from setting the fee down to zero if they want, by setting the fee value to an actual point. But I'd make the default be [0, N] where N is some positive value/percentage. The current case can be imperfectly mapped into this design by setting the starting fee to this N, the top value in the allowed range (with the exception that, in the current scheme, once you decrease the fee, you cannot increase it back -- right?).

**Comment by augustoteixeira** [comment]
I really like this idea. Users would see two variables when selecting pools:
- current commission
- maximum commission

Then the pool manager is able to tune the commission at will respecting 
the established maximum.

**Comment by tuler** [comment]
> The current case can be imperfectly mapped into this design by setting the starting fee to this N, the top value in the allowed range (with the exception that, in the current scheme, once you decrease the fee, you cannot increase it back -- right?).

Right. It's a simple change to allow increase back to the initial value.
The simplest option is to use the initial value as upper bound. Or make it two parameters, `initial` and `max`, just to avoid an additional transaction to set the `initial` upper bound to the desired amount.

For simplicity I'd required the additional transaction. It's cleaner and will be part of the commission value history.

**Comment by tuler** [comment]
There is a change though that operators will create pools with 100% commission just to have full flexibility.

**Comment by packetstracer** [comment]
> > There is a change though that operators will create pools with 100% commission just to have full flexibility.
> 
> Maybe there needs to be a "maximum maximum"? I haven't done any serious surveys, but I cannot conceive of any running pool of any crypto protocol out there that would be operating on a fee higher than, say, 10% of the block's value, and whatever the equivalent would be in the gas/oracle fee model.
> 
> Maybe the pool operator community could chime in here and discuss a possible "maximum maximum fee"?

It's not that strange to see commissions >= 10% (for example in Elrond network and Fantom it's fixed at like 15-17%, in cosmos networks it's set usually between 5-10%).

Also have in mind:
- what could be the APR of staking in the future: for example if the APR goes to like 10%, that would be a 1% maximum income for a validator
- the costs that a validator will have to afford: if for example, new hardware requirements are set (NVM's SSDs, more cores, RAM whatever, minimum uptimes, ...)

I think that a maximum of 20% commission would be a more "safe" amount to enforce.

**Comment by packetstracer** [comment]
> It is important to understand two aspects here:
> 
> * Cartesi's commission system already accounts for increases in Ethereum fees. If said fees go high, then the commission increases as well
> * It has always been possible to decrease the commission, what is at question is whether it should be possible to increase them.
> * Note that some users may subscribe to the pool and never look at this again. So if a pool increases its commission, the users may be loosing money without realizing.
> * Also it is important to observe that the users pay a reasonable cost to swap between two different pools. So if a pool with lots of users decides to increase its commission, the users may become "trapped" in a situation that is not clear whether it is worth paying the transaction cost of changing pools to look for some other group (which can also increase the fees later on).

My 2 cents about the 2 last points:

- it's fair to protect delegators but...
- it's delegator's responsability to investigate and check whom they're delegating to and what the conditions are (my bank account conditions may change with time, I'm the one to decide upon that)
- I agree with that Eth fees may be a stop to redelegating to another pool, but It seems like Ethereum will fix/relieve that with ZK/Optimism rollups and PoS in the "near" future. Also if I was a delegator I wouldn't hesitate to pay 20$ to redelegate if my pool has set the fee's let's say to 50%, but for sure would be kinda hard to make up a decision if the commission varies in a quite narrow range

Thinking as a validator, If I cannot increase the commission I would set a higher one in order to not fall into a risky stage of not being profitable enough and not being able to revert back. I would also not be able to set up very low commissions for a limited time to attract more delegators.

In the end if other delegators think in this manner, the average overall commission in the network will end up being higher and thus the delegators earning less rewards. 

I agree with the maximum fee and maximum fee change, but not with not being able to increase the fee, I think it's bad for the validators and also for delegators.

**Comment by packetstracer** [comment]
You can check different Cosmos networks in this block explorer, just select the network in the header select input and click on the validator tab.

[https://explorer.elrond.com/providers](https://www.mintscan.io/cosmos/validators)

this one for the Elrond network

[https://explorer.elrond.com/providers](https://explorer.elrond.com/providers)

**Comment by fcecin** [comment]
> * what could be the APR of staking in the future: for example if the APR goes to like 10%, that would be a 1% maximum income for a validator

The pool operator does not lock capital, right? APR on locked capital on a pool seems like a bad heuristic measure for discussing the level of funding that a pool operator needs. 

> * the costs that a validator will have to afford: if for example, new hardware requirements are set (NVM's SSDs, more cores, more RAM, minimum uptimes, whatever...)

The amount of actual trade value that a pool operator needs is determined by a free competition between pool operators. No pool operator can demand unreasonable actual trade value fees and survive because they will be out-competed by another pool operator that demands less fees. That's what we expect.

So, the job of a "maximum maximum" fee is not really to answer questions about the technical viability of pool operators vs. e.g. the trade value of CTSI tokens and the amount of CTSI awarded by mining a block -- that is determined by the competition between pool operators -- but to prevent abuse from pool operators that would ever set the maximum fee to 100% "preventively", which is absolutely ridiculous on multiple counts and a clear abuse of the system and possibly even a way to troll the Noether network, even if these statically, objectively outrageous fee levels are never actually used.

If pool operators are legitimately being disabled by the inability to set a fee to something like 40%, which is objectively ludicrous regardless of the actual value produced by a block, then the fee % (or whatever the equivalent in the gas-oracle model) is the wrong tool to solve that issue. Then maybe the Noether protocol should just be forked to produce more CTSI, or maybe we should be OK with US$100/month simple cloud nodes producing blocks instead of some US$10,000/month bare-metal or cloud hosting boutique doing it, or maybe something other solution entirely.

What level of block rewards fee for pool operators starts to look like a technical failure, purely from a numerical standpoint?

I think something like 20% is really stinky, but maybe that's just my opinion. I don't think that any pool operator would or should survive asking for 20% fees, when, in normal conditions, they should have been long out-competed by the other pool operators. I still think 10% would be better, but that's just my opinion. I actually should not be debating this because I'm the CIP Editor (a position that has meta powers) so I'm certainly not going to be voting on this.

**Comment by tuler** [comment]
I don't like the "maximum maximum" idea. It's a difficult task to come up with this value.

**Comment by fcecin** [comment]
> I don't like the "maximum maximum" idea. It's a difficult task to come up with this value.

We can also just tell users to be completely responsible when selecting pools, and observe the maximum fee of the pool carefully.

**Comment by tuler** [comment]
Don't you think pools will just have a maximum of 100% for the sake of flexibility?
Maybe we can't run from a cool-down period (i.e. raise only once a week).

Not my favorite solution, but it can be a simple implementation.
Set to whatever value you want, but if you raise you can do it again only after some period of time.

**Comment by fcecin** [comment]
> Don't you think pools will just have a maximum of 100% for the sake of flexibility?
> Maybe we can't run from a cool-down period (i.e. raise only once a week).
> 
> Not my favorite solution, but it can be a simple implementation.
> Set to whatever value you want, but if you raise you can do it again only after some period of time.

If the "maximum fee" parameter does not exist at all, then I would agree with this. But if there is an explicit "maximum fee" parameter, and some can set it to 10% and others to 100%, and people started creating this narrative that the 100% maximum parameter is "necessary", that would start to really stink. So I'd either enable a "maximum fee" parameter with a hardcoded "maximum maximum" of 10% to 20%, or I would just enable pool operators to set their fees to whatever.

**Comment by packetstracer** [comment]
Here you are a cosmos network validator screenshot
![image](https://user-images.githubusercontent.com/8169192/127193738-ea2a7b12-9a94-402a-bdee-e0bba1b548ed.png)

- Commision rate has a cooldown value of 1 day 
- Max rate is the maximum set by the validator only at creation time
- Max change rate is the maximum change to that max rate
- Bonding and redelegation periods are 21 days

Though this values are freely set by the validator, it's quite hard to see a validator with a comission greater that 20% in any network (and in almost any cosmos network is <=10%). 

Why are not every validator setting the fee to 30-50-100%? Mainly because they would loose their delegations and would just disappear. The validators fees get regulated by "free market" competition and that seems to work for Cosmos networks and others such as Cardano (I'm not saying this is the best approach IMO but seems to work well even though the redelegation period is like 10 times greater than Cartesi's).

"Don't you think pools will just have a maximum of 100% for the sake of flexibility?" 

Usually not, because the delegator can take this into account when selecting the delegator (He may wonder, Won't this validator set a 100% fee from overnight?). In my case, I would never set that value greater than 15-20%.


Anyway, that's up to you...going completely free market, a constrained  model or something in between. 

Let's see if other validators express/develop their opinions on this.

**Comment by tuler** [comment]
Thank you for the feedback and the example.

> * Max change rate is the maximum change to that max rate

You mean "maximum change to the current value, right? Max rate is fixed at creation time as you said.

In the example, does 10% max change rate means it can go from 10% to 11%, or from 10% to 20%?

**Comment by packetstracer** [comment]
> Thank you for the feedback and the example.
> 
> > * Max change rate is the maximum change to that max rate
> 
> You mean "maximum change to the current value, right? Max rate is fixed at creation time as you said.

Yes, change from the current value. 

> 
> In the example, does 10% max change rate means it can go from 10% to 11%, or from 10% to 20%?

A 10% max change allows you to set from 10% to 20% or to 0% when editing your validator details. 

I think "Max change rate" is also set only at creation time.

**Comment by 0rtis** [comment]
I suggest to implement a target based fee that is modifiable anytime by the pool operator. Every period of time `T` (`T` being a number of block or a fixed duration), the pool fee `F` is updated by a fixed amount `Fd` toward the target `Ft`. `Fd` is fixed and set by the protocol. After each update, the new fee `F` is equal to `max(Ft, F - Fd)` if `Ft < F` and `min(Ft, F + Fd)` if `Ft > F`

Example:
A pool is deployed with an initial fee `F` of 10%. After a while, the pool operator decided to decrease the fee in order to attract more delegator and set the target `Ft` to 5%. After `T` as passed, the fee is updated so that the new fee is equal to `F - Fd`. Assuming `Fd` is 1%, the new fee `F` is 9%. `F` will decrease every `T` period until it reaches the target `Ft` set to 5%.

At anytime, even while the `F` is moving to a new target `Ft`, the pool operator can change the target `Ft`. Continuing with our example, after the first update from 10% to 9%, the pool operator may change his/her mind and decide to keep the fee to  10%. He/she update the target fee to 10%. On the next update (after another `T` has passed), the fee will now increase by 1% (`F + Fd`) and the new fee will be 10%.

This setup has the advantage to allow flexible management of the fee based on market condition while protecting the delegators from sudden increase in fee.

**Comment by augustoteixeira** [comment]
Here is another suggestion:
- The pool manager is able to modify the fees anyway they want.
- But when doing so, there is a cool-down period of a few weeks before the new fees enter in effect. This is imposed so that the users have time to get informed about the change and decide whether they want to stake somewhere else.

This could be a good trade-off between pool users and pool managers.

**Comment by packetstracer** [comment]
> Here is another suggestion:
> 
> * The pool manager is able to modify the fees anyway they want.
> * But when doing so, there is a cool-down period of a few weeks before the new fees enter in effect. This is imposed so that the users have time to get informed about the change and decide whether they want to stake somewhere else.
> 
> This could be a good trade-off between pool users and pool managers.

I think this is a simple and nice approach. But as a validator would prefer a not so long cool-down period...though it's not a great issue but it's nice to react fast to other validator's changes. If it takes a month or two to change the fee could be too long.

I think 1 week could be nice having in mind that bonding period of the delegation is quite low. So a user would have a whole week to redelegate.

Also it could be applied only in the increase fee direction, allow a fee decrease sending a notice to the delegator. A decrease in fees is always a good thing for the delegator, so could be good for both parts, allowing instant fee decrease and cooled-down fee increase.

**Comment by fcecin** [comment]
> Here is another suggestion:
> 
> * The pool manager is able to modify the fees anyway they want.
> * But when doing so, there is a cool-down period of a few weeks before the new fees enter in effect. This is imposed so that the users have time to get informed about the change and decide whether they want to stake somewhere else.
> 
> This could be a good trade-off between pool users and pool managers.

During that time, the fee collected by the pool could also be reduced by some amount. That would (i) create a disincentive for pools to increase their fees and (ii) it would give dissatisfied users a discount for the set period, and when that period nears its end, whatever cost they would have to bear to switch to another pool would have been amortized by the fee discount. The discount can be steep (even something like 50%), since it is temporary anyway. This could also help keep the fee change delay to a minimum (e.g. 1 week).

**Comment by augustoteixeira** [comment]
Concerning the decision on the exact value of the cool-off period: this 
could perhaps be chosen by the pool managers themselves during pool 
creation (and then stay fixed forever).

Then, this period would be another factor that the users would take in 
consideration when choosing the pool.

**Comment by tuler** [comment]
In my opinion choosing a pool taking this cool-off period into consideration is strange. Imagine how that information will be presented in the explorer. Remember there is no notification mechanism involved. With no max change limits, allowing it to go from 1% to 80%, even with the cool-off is very strange.

Taking the thread comments into consideration I like the following design:

```
1) the pool manager creates a pool with an initial fee (i.e. 8%).
2) the pool manager can lower the fee any time, effective immediately.
3) the pool manager can raise the fee by a maximum of 5 pp. (i.e. from 8% to 13%), effective immediately.
4) once raised he can only raise again after 1 week.

5 pp. is imposed by the protocol.
1 week is imposed by the protocol.

Same design applies to gas tax based fee. By replacing percentage with gas quantity. Instead of 5 pp, it will be i.e. 10000 gas.
```

**Comment by augustoteixeira** [comment]
I like this proposal as well. It gives people time to consider alternatives.

**Comment by cf-cartesi** [comment]
Guess we have converged for a v1. A fixed maximum % (i.e. 5%) or gas amount (i.e. 10 000) of variation with a certain cool-down (i.e. 1 week) when increasing the commission and no restrictions to decrease it. That seems to fit the demand of pool owners to be able to increase the rate while protecting users from sudden prohibitive commission increases. The example values provided also make good candidates for the v1 parameters (5% max increase/week for flat rate commissions and 10 000 gas increase/week for gas-based commissions), IMHO. Any showstoppers in your opinions @packetstracer @0rtis @vaddec-everstake?

**Comment by packetstracer** [comment]
> Guess we have converged for a v1. A fixed maximum % (i.e. 5%) or gas amount (i.e. 10 000) of variation with a certain cool-down (i.e. 1 week) when increasing the commission and no restrictions to decrease it. That seems to fit the demand of pool owners to be able to increase the rate while protecting users from sudden prohibitive commission increases. The example values provided also make good candidates for the v1 parameters (5% max increase/week for flat rate commissions and 10 000 gas increase/week for gas-based commissions), IMHO. Any showstoppers in your opinions @packetstracer @0rtis @vaddec-everstake?

Looks Good To Me

It's fair enough for us validators and delegators. The stopper for me was being unable to increase the fee and have to move to another pool.
